chainer/functions/loss/ctc.py

```diff
@@ -0,0 +1,247 @@
+        l_type = in_types[0]
+        utils.type_check.expect(l_type.dtype == numpy.int32)
+
+        x_basetype = in_types[1]
+
+        for i in range(2, len(in_types)):
+            x_type = in_types[1]
+            utils.type_check.expect(
+                x_type.dtype == numpy.float32,
+                x_type.shape == x_basetype.shape,
+            )
+
+    def log_matrix(self, x):
+        xp = cuda.get_array_module(x)
+        if xp == numpy:
+            res = numpy.ma.log(x).filled(fill_value=self.zero_padding)
+        else:
+            create_recurrence_relation = cuda.cupy.ElementwiseKernel(
+                'T x, T e', 'T y',
+                '''
+                if(x == 0){
+                      y = e;
+                }else{
+                      y = log(x);
+                }
+                ''',
+                'create_recurrence_relation')
+            res = create_recurrence_relation(x, self.zero_padding)
+        return res
+
+    '''
+    Transition in forword and backword algorithms is represented as matrix.
+    See also
+    https://blog.wtf.sg/2014/10/06/connectionist-temporal-classification-ctc-with-theano/
+    '''
+    def recurrence_relation(self, size, xp):
+        rr = (xp.eye(size) + xp.eye(size, k=1) +
+              xp.eye(size, k=2) * (xp.arange(size) % 2))
+        return self.log_matrix(rr)
+
+    def label_to_path(self, labels):
+        xp = cuda.get_array_module(labels)
+        label_length = labels.shape[0]
+        path = xp.full((label_length * 2 + 1,),
+                       self.blank_symbol, dtype=int)
+        for i in range(label_length):
+            path[i * 2 + 1] = labels[i]
+        return path
+
+    def log_dot(self, prob, rr):
+        xp = cuda.get_array_module(prob)
+        res = xp.zeros(prob.shape)
+        rtrans = xp.swapaxes(rr, 1, 0)
+        for i in range(rtrans.shape[0]):
+            res[i] = self.logsumexp(prob + rtrans[i])
+        return res
+
+    def logsumexp(self, a):
+        xp = cuda.get_array_module(a)
+        vmax = xp.amax(a)
+        res = xp.log(xp.sum(xp.exp(a - vmax)))
+        return (res + vmax).astype(numpy.float32)
+
+    # path probablity to label probability
+    def label_probability(self, label_size, path, multiply):
+        xp = cuda.get_array_module(path)
+        chars = set([c for c in path])
+        labels_prob = self.log_matrix(xp.zeros((label_size,),
+                                               dtype=numpy.float32))
+        if xp == numpy:
+            for c in chars:
+                pos = numpy.where(path == c)[0]
+                labels_prob[c] = self.logsumexp(xp.take(multiply, pos))
+        else:
+            cuda.cupy.ElementwiseKernel(
+                'raw float32 x, raw I y, I s',
+                'float32 z',
+                '''
+                float value = z;
+                for(int index=0;index<s;++index){
+                    if(y[index] == i){
+                        if(value > x[index]){
```

Comment:
Index access is high cost in cuda kernel.
It is better to be assigned to a temporary variable.